utils/fixture.js

```diff
@@ -1,99 +1,102 @@
-  fixture = fixture && fixture[name] ? fixture[name] : undefined
+function setFixture (mock, fixture, name, selectedType, schema) {
+  fixture = fixture !== undefined ? fixture[name] : undefined
 
   if (fixture === undefined) {
     return mock
   }
 
-  if (Array.isArray(fixture)) {
-    return fixture.map((val) => {
-      if (isObject(val)) {
-        return Object.assign({}, handleObjectFixture(mock[0], val, name))
-      }
+  return validateFixture(mock, fixture, selectedType, schema, name)
+}
 
-      return scalarValidation(mock[0], val, name)
-    })
-  } else if (isObject(fixture)) {
-    return handleObjectFixture(mock, fixture, name)
-  } else {
-    return scalarValidation(mock, fixture, name)
+function validateFixture (mock, fixture, selectedType, schema, name) {
+  if (selectedType.noNull && fixture === null) {
+    throw new Error(`${selectedType.name} can't be null.`)
   }
-}
 
-function scalarValidation (mock, fixture, name) {
   if (fixture === null) {
     return null
   }
 
-  if (Array.isArray(mock) && !Array.isArray(fixture)) {
-    throw new Error(`${name} fixture is not an array and it should be one.`)
+  if (selectedType.isArray && !Array.isArray(fixture)) {
+    throw new Error(`${selectedType.name} fixture is not an array and it should be one.`)
   }
 
-  if (typeof mock !== typeof fixture) {
-    throw new Error(`${name} fixture is not the same type as the document.`)
-  }
+  if (schema[selectedType.type]) {
+    const fields = schema[selectedType.type].fields
 
-  return fixture !== undefined ? fixture : mock
-}
+    if (Array.isArray(fixture)) {
+      return fixture.map(val => {
+        return validateFixture(mock[0], val, schema[selectedType.type], schema, name)
+      })
+    } else if (isObject(fixture)) {
+      for (const val of Object.keys(fixture)) {
+        const selectedField = fields.filter(field => field.name === val)
+        // If it's extra fields on the fixture, continue
+        if (!selectedField.length) {
+          continue
+        }
 
-function handleObjectFixture (mock, fixture, name) {
-  for (const val of Object.keys(fixture)) {
-    // If the value is defined on the fixture but not on the query, skip that value
-    if (typeof mock[val] === 'undefined') {
-      continue
-    }
+        // If it is a custom scalar, should not validate the typeof
+        if (isObject(mock[val]) && Object.keys(mock[val]).length === 1 && mock[val].__typename) {
+          mock[val] = fixture[val]
+        } else {
+          mock[val] = validateFixture(mock[val], fixture[val], selectedField[0], schema, name)
+        }
+      }
 
-    // If it is a scalar, it should be an empty object instead of the __typename
-    if (isObject(mock[val]) && Object.keys(mock[val]).length === 1 && mock[val].__typename) {
-      mock[val] = {}
+      return mock
     }
+  } else if (Array.isArray(fixture)) {
+    fixture.forEach(val => {
+      if (selectedType.isArray && selectedType.noNullArrayValues && val === null) {
```

Comment:
should not be this validation also moved inside `validateType`?

Reply:
Yes! much better!! Thanks!